### ISSUE TITLE
New appliance format (version 8)

### ISSUE THREAD

**Issue by grossmj** [GNS3/gns3-registry#734]
### Discussed in https://github.com/GNS3/gns3-registry/discussions/729

<div type='discussions-op-text'>

<sup>Originally posted by **grossmj** January 16, 2023</sup>
With GNS3 v3.0 around the corner, it would be a great opportunity to change the appliance file format to add flexibility for custom settings / hardware configuration for specific versions and also add some other improvements:

Here is what I have in mind for now:

* Support for more than one emulator settings / hardware configurations. One config can be assigned to a specific version using its name. The default config is used (`default` field must be configured if there are multiple configs) or the only config present if none is explicitly assigned to a version. Should we refer to this as "configs" or "settings"?
* Support for different checksum algorithms by adding a `checksum_type` field (md5, sha256 etc.) and replacing `md5sum` by a `checksum` field in the `images` section. What  algorithms should we support in addition of md5 and sha256?
* Should we add `default_username` and `default_password` fields?

Example:

```
{
    "appliance_id": "709c2a9b-5dc3-4362-b147-fb848a0df963",
    "name": "My appliance",
    "category": "router",
    "description": "This is my new appliance",
    "vendor_name": "Cisco",
    "vendor_url": "http://www.cisco.com/",
    "documentation_url": "https://www.cisco.com/c/en/us/support/routing/xxx",
    "product_name": "Appliance product xxx",
    "product_url": "https://www.cisco.com/c/en/us/products/xxx/index.html",
    "registry_version":8,
    "status": "experimental",
    "maintainer": "GNS3 Team",
    "maintainer_email": "developers@gns3.net",
    "usage": "This is how to use my appliance",
    "default_username": "cisco",
    "default_password": "admin",
    "symbol": "router.svg",
    "first_port_name": "eth0",
    "port_name_format": "eth{port1}",
    "settings": [
        {
            "name": "Default settings",
            "type": "qemu",
            "default": true,
            "adapter_type": "e1000",
            "adapters": 2,
            "ram": 4096,
            "cpus": 1,
            "hda_disk_interface": "scsi",
            "arch": "x86_64",
            "console_type": "vnc",
            "boot_priority": "cd",
            "kvm": "require",
            "options": ""
        },
        {
            "name": "Custom settings for version 7.10.2",
            "type": "qemu",
            "adapters": 4,
            "ram": 8192,
            "cpus": 1
        }
    ],
    "images": [
        {
            "filename": "file.iso",
            "version": "7.10.2",
            "checksum": "ef8712e655fcbc92dc1a1551ee2e4a80",
            "checksum_type": "md5",
            "filesize": 1287245824,
            "download_url": "https://software.cisco.com/download/home/286307342/type/286307754/release/7.10.2"
        },
        {
            "filename": "file2.iso",
            "version": "6.10.4",
            "checksum": "e5b844cc57f57094ea4585e235f36c78c1cd222262bb89d53c94dcb4d6b3e55d",
            "checksum_type": "sha256",
            "filesize": 3949459594,
            "download_url": "https://software.cisco.com/download/home/286307342/type/286307754/release/6.10.4"
        },
        {
            "filename": "empty100G.qcow2",
            "version": "1.0",
            "checksum": "1e6409a4523ada212dea2ebc50e50a65",
            "checksum_type": "md5",
            "filesize": 198656,
            "download_url": "https://sourceforge.net/projects/gns-3/files/Empty%20Qemu%20disk/",
            "direct_download_url": "https://sourceforge.net/projects/gns-3/files/Empty%20Qemu%20disk/empty100G.qcow2/download"
        }
    ],
    "versions": [
        {
            "name": "7.10.2",
            "settings": "Custom settings for version 7.10.2",
            "images": {
                "hda_disk_image": "empty100G.qcow2",
                "cdrom_image": "file.iso"
            }
        },
        {
            "name": "6.10.4",
            "images": {
                "hda_disk_image": "empty100G.qcow2",
                "cdrom_image": "file2.iso"
            }
        }
    ]
}

```

Please let me know if you have comments / suggestions.</div>

**Comment by josephmhiggins** [comment]
1) The way I provide support for Links in my powershell code is to be abstract, e.g. in the Gns3 community forum Backup-AppDataGns3Ver22NoProfiles.ps1 and How to Use Powershell to Handle gns3server log Creation in Gns3 2.2.34, is vastly superior than providing direct links in the .gnsa files. My Links find the webpages. Companies are always moving webpages around. The .gnsa files should not provide direct links because direct links are only valid for ....7 days.
2) a username a separate password field in the description sounds good to me, except I would change it to username_default and password_default. But people get testy over variable names on the internet, so I really do not care.

**Comment by eantowne** [comment]
>Should we refer to this as "configs" or "settings"?
- I think "settings" may be better so as to not cause confusion with the configuration within the given OS.
>What algorithms should we support in addition of md5 and sha256?
- Honestly I think that, from a security perspective to prevent/reduce the chance of a malicious actor taking advantage of a collision, SHA256 would be the minimum. Most vendors are providing SHA256 and it seems to be the general standard. I truly do not see a reason for providing multiples, beyond recursive validation by a user for manual download to compare the hash in the .gns3a/Marketplace vs the vendor provided hash.
>Should we add default_username and default_password fields?
- I think this is a glorious idea!

**Comment by grossmj** [comment]
Thanks @eantowne, I will also use your ideas from the discussion to create the new format. Also, I think we should just make it a evolution from the current format (aka version 8) and still support older versions in GNS3 v3.0 otherwise it would be a nightmare to maintain two different branches for appliances.

**Comment by eantowne** [comment]
I would agree with enabling backwards compatibility. I would use version 30 though, gives plenty of room for updates tot he 2.x branch, and clearly delineates itself as the new and improved  contestant.

One thing does occur to me though, and that is how is this going to work for the Marketplace? For instance, my intention is to create a whole bunch of updates for a wide variety of appliances to take advantage of the  new file capabilities. This will basically create duplicate .gns3a's for certain appliances. Does there need to be some form of check by the Marketplace sync to look if the .gns3a is 3.0 compatible and then label it as such? Or would it be easier to have two separate appliance list in the Marketplace? Even then, a 3.0 controller will be able to read and import a v4 or 3.0 compatible, bu then when do we stop updating 2.x compatible appliances?

**Comment by grossmj** [comment]
> One thing does occur to me though, and that is, how is this going to work for the Marketplace? For instance, my intention is to create a whole bunch of updates for a wide variety of appliances to take advantage of the new file capabilities. This will basically create duplicate .gns3a's for certain appliances. Does there need to be some form of check by the Marketplace sync to look if the .gns3a is 3.0 compatible and then label it as such? Or would it be easier to have two separate appliance list in the Marketplace? Even then, a 3.0 controller will be able to read and import a v4 or 3.0 compatible, but then when do we stop updating 2.x compatible appliances?

Actually I was thinking we would update appliances to take advantage of the new file capabilities only when needed, for instance if we need different hardware configurations. I know this is not ideal and that would mean these appliances cannot be used in GNS3 v2.x

I think there are 2 possible solutions:

- Copy all .gns3a appliances to a new directory in the repo and convert them to the new format. The marketplace would then have two separate appliance lists with a toggle button to switch between v2 and v3. Main problem is we would have to maintain 2 separate lists for a while until v3 is well established (this could be for a long time, especially I predict some users will want to stay on v2 because the local server is supported on Windows).

OR

- We keep the current .gns3a files, upgrade the ones that should use the new capabilities and support the new format in GNS3 >= v2.2.38. This would involve some major work from us and and also force everyone who want to use appliances with the new format to upgrade their GNS3 version.

**Comment by grossmj** [comment]
> Copy all .gns3a appliances to a new directory in the repo and convert them to the new format. The marketplace would then have two separate appliance lists with a toggle button to switch between v2 and v3. Main problem is we would have to maintain 2 separate lists for a while until v3 is well established (this could be for a long time, especially I predict some users will want to stay on v2 because the local server is supported on Windows).

We could also have 2 different branches, `master` for v2 and `3.0` for v3. Then we would just merge changes done on master into 3.0

**Comment by eantowne** [comment]
>We could also have 2 different branches, master for v2 and 3.0 for v3. Then we would just merge changes done on master into 3.0

Wouldn't this merge be a massive issue?

For instance Juniper vMX current .gns3a does not support the ability to have a multi-RE/PFE configuration. The new format would. How does the 2.x format merge into the 3.0 format without causing issues, or at least confusion if going with a 2 file model.

I think there may be an argument for just keeping the 2.x and 3.x branches completely separate. If the 2.x branch does not get "official" updates, but the 3.x does, this might also help push people to upgrade to 3.x if they cannot find the template they need in 2.x, but it is sitting there in 3.x.

There comes a point where supporting the "old" versions may not be worth the effort. I am not saying just immediately stop all 2.x registry work the day that 3.x is released, but I think a planned obsolesce might be a good thing. Announcing an end-date for 2.x registry support projected out X months from 3.x release?

**Comment by grossmj** [comment]
> Wouldn't this merge be a massive issue?

Correct, this is why will we introduce a version 8 of the current format and support it in both >= v2.2.41 and 3.0. This means a bit more work for us but I believe this will create less headaches down the road.

**Comment by grossmj** [comment]
> Wouldn't this merge be a massive issue?

Correct, this is why will we introduce a version 8 of the current format and support it in both >= v2.2.41 and 3.0. This means a bit more work for us but I believe this will create less headaches down the road.

**Comment by grossmj** [comment]
I have tried to include all the suggestions but please let me know if you see something that needs to be added/changed. Support for more that 4 drives for Qemu VMs will be added later in 3.x so that should be added in another version of the appliance format.

Summary of what this version will support:

* uefi boot mode option for Qemu
* Possibility to have multiple set of emulator settings to use with different image versions. Default settings are used if a `default` field set to `true` exists or there is only one set present. An `emulator_type` field must be added to tell GNS3 what emulator to use (`qemu`, `iou` or `dynamips`), this could allow one appliance to be used with different emulators. All emulator specific properties are defined in a `emulator_properties` field.
* New optional `emulator_versions` field in the settings to restrict using the emulator properties on specific emulator versions.
* New optional `checksum_type` field for future development. The default and only checksum type remains MD5 for now.
* New optional `default_username` and `default_password` fields.
* New optional `installation_instructions` field to give download/unpack instructions to some appliances. See https://github.com/GNS3/gns3-registry/issues/506 
* New optional `compression_target` to be used along with the `compression` field in future development.
* The `usage`, `default_username`, `default_password`, `symbol`, `first_port_name`, `port_name_format` fields can be defined at the "version" level. Defaults are at the appliance level.

Example:

```
{
    "appliance_id": "709c2a9b-5dc3-4362-b147-fb848a0df963",
    "name": "My appliance",
    "category": "router",
    "description": "This is my new appliance",
    "vendor_name": "Cisco",
    "vendor_url": "http://www.cisco.com/",
    "documentation_url": "https://www.cisco.com/c/en/us/support/routing/xxx",
    "product_name": "Appliance product xxx",
    "product_url": "https://www.cisco.com/c/en/us/products/xxx/index.html",
    "registry_version":8,
    "status": "experimental",
    "maintainer": "GNS3 Team",
    "maintainer_email": "developers@gns3.net",
    "usage": "This is how to use my appliance",
    "default_username": "cisco",
    "default_password": "admin",
    "symbol": "router.svg",
    "first_port_name": "eth0",
    "port_name_format": "eth{port1}",
    "settings": [
        {
            "name": "Default settings",
            "default": true,
            "emulator_type": "qemu",
            "emulator_properties":
                {
                    "adapter_type": "e1000",
                    "adapters": 2,
                    "ram": 4096,
                    "cpus": 1,
                    "hda_disk_interface": "scsi",
                    "arch": "x86_64",
                    "console_type": "vnc",
                    "boot_priority": "cd",
                    "kvm": "require",
                    "options": ""
                }
        },
        {
            "name": "Custom settings for version 7.10.2",
            "emulator_type": "qemu",
            "emulator_versions": ["6.2.0"]
            "emulator_properties":
                {
                    "adapters": 4,
                    "ram": 8192,
                    "cpus": 1
                }
        }
    ],
    "images": [
        {
            "filename": "file.iso",
            "version": "7.10.2",
            "checksum": "ef8712e655fcbc92dc1a1551ee2e4a80",
            "checksum_type": "md5",
            "filesize": 1287245824,
            "download_url": "https://software.cisco.com/download/home/286307342/type/286307754/release/7.10.2"
        },
        {
            "filename": "file2.iso",
            "version": "6.10.4",
            "checksum": "e5b844cc57f57094ea4585e235f36c78c1cd222262bb89d53c94dcb4d6b3e55d",
            "checksum_type": "md5",
            "filesize": 3949459594,
            "download_url": "https://software.cisco.com/download/home/286307342/type/286307754/release/6.10.4"
        },
        {
            "filename": "empty100G.qcow2",
            "version": "1.0",
            "checksum": "1e6409a4523ada212dea2ebc50e50a65",
            "checksum_type": "md5",
            "filesize": 198656,
            "download_url": "https://sourceforge.net/projects/gns-3/files/Empty%20Qemu%20disk/",
            "direct_download_url": "https://sourceforge.net/projects/gns-3/files/Empty%20Qemu%20disk/empty100G.qcow2/download"
        }
    ],
    "versions": [
        {
            "name": "7.10.2",
            "settings": "Custom settings for version 7.10.2",
            "usage": "Specific usage instruction for version 7.10.2",
            "default_username": "admin",
            "default_password": "admin",
            "symbol": "multilayer_router.svg",
            "first_port_name": "ethernet0",
            "port_name_format": "ethernet{port1}",
            "images": {
                "hda_disk_image": "empty100G.qcow2",
                "cdrom_image": "file.iso"
            }
        },
        {
            "name": "6.10.4",
            "images": {
                "hda_disk_image": "empty100G.qcow2",
                "cdrom_image": "file2.iso"
            }
        }
    ]
}

```

**Comment by yueguobin** [comment]
# Appliance format v8: spec vs gns3-server implementation status

Following up on the v8 spec summarized above (June 2023), here is a field-by-field comparison against what gns3-server actually implements on the `3.1` branch, split into **schema** (parsing / listing) and **install** (appliance → template conversion). The last column is the v8 install support implemented on top of 3.1 (work in progress, PR to follow).

## Observed behavior on stock 3.1

A v8 appliance (currently `home-assistant.gns3a` and `vyos.gns3a`) **parses and lists fine** — the `ApplianceV8` pydantic model behind the `registry_version` discriminated union handles it — but **installing fails**:

- `ApplianceToTemplate.new_template()` (`gns3server/controller/appliance_to_template.py`) only dispatches on the v1-6 top-level `qemu` / `iou` / `dynamips` / `docker` blocks. A v8 file has none of them, so no branch matches and the resulting template dict contains neither `template_type` nor any emulator property → rejected by `TemplateCreate` with `template_type Field required`.
- Reproduced with the current `home-assistant.gns3a`: the conversion output is only `{compute_id, name, version, usage, category, symbol}`.
- Extra pain point: image download runs before the conversion, so a missing image (1 GB for haos) is downloaded first and the failure only comes after.

## A. Top-level fields

| Spec item | 3.1 (schema) | v8 install (WIP) |
|---|---|---|
| `registry_version: 8` discriminator, old versions still supported | ✅ discriminated union on `registry_version` (`ApplianceV1_6` / `ApplianceV8`) | ✅ |
| `default_username` / `default_password` (appliance level) | ✅ | ✅ kept on the template in a new `appliance_metadata` JSON column and seeded onto nodes |
| `installation_instructions` | ✅ | ✅ kept in appliance metadata |
| `vendor_url` and `maintainer_email` required | ✅ | ✅ |
| v1-6 top-level `qemu`/`iou`/`dynamips`/`docker` blocks replaced by `settings` | ✅ `settings: List[TemplateSetting]` (required) | ✅ |
| v1-6 top-level `first_port_name`, `port_name_format`, `port_segment_size`, `linked_clone` removed (moved into qemu properties) | ✅ removed | ✅ |
| *(beyond spec)* `vendor_logo_url` | ✅ (marketplace) | ✅ |
| *(beyond spec)* `netmiko_device_type` (device type for Netmiko-based automation) | ❌ | ✅ added at appliance level (v1-6 and v8), inherited by templates/nodes |

## B. settings / template_properties

| Spec item | 3.1 (schema) | v8 install (WIP) |
|---|---|---|
| Multiple settings sets; default = `default: true` set, or the only set present | ✅ fields | ✅ same three rules: version-referenced set > `default: true` > single set |
| Version references a settings set by `name` | ✅ `ApplianceVersionV8.settings` | ✅ errors with the missing name when unresolvable |
| `template_type` (`qemu`/`iou`/`dynamips`/`docker`); one appliance can create different template types | ✅ union of per-type property models | ✅ |
| Default settings inherited; `inherit_default_properties: false` blocks it | ✅ field (default `true`) | ✅ implemented; additionally only inherits from a default set of the *same* template type (not spelled out in the spec) |
| Undefined property falls back to the controller template default | — | ✅ absent properties simply go through `TemplateCreate` defaults |
| `name` + `default_name_format` in all template properties | ✅ | ✅ |
| `category` / `usage` / `symbol` definable in `template_properties`, injected from version/appliance level when absent | ✅ fields | ✅ cascade: properties > version > appliance |
| qemu: `arch` renamed `platform` | ✅ | ✅ |
| qemu: `first_port_name`, `port_name_format`, `port_segment_size`, `linked_clone`, `custom_adapters` in properties | ✅ | ✅ |
| qemu: `uefi` | ✅ (plus `tpm`, beyond spec) | ✅ |
| qemu: `spice+agent` console type | ✅ | ✅ |
| **qemu: `kvm` dropped, no longer required** | ❌ **spec violation**: `QemuPropertiesV8.kvm` is still `Field(...)` (required) — a spec-compliant v8 file that omits `kvm` (e.g. the current `home-assistant.gns3a`) is rejected at validation time | ✅ made optional; `kvm: disable` is translated to `-machine accel=tcg` in qemu options like the v1-6 path |
| dynamips: `idlepc` moved into template properties | ✅ present in `DynamipsPropertiesV8` | ✅ properties take precedence, with a lenient version-level fallback (beyond spec) |
| docker: `console_resolution`, `extra_hosts`, `extra_volumes` | ✅ | ✅ |
| *(beyond spec)* docker: `custom_adapters` (port-named Docker vendor appliances) | ❌ | ✅ added |
| *(beyond spec)* docker: `extra_configs` (config files bind-mounted read-only into the container) | ❌ | ✅ added |
| *(beyond spec)* `cpu_throttling` | float 0–100, inconsistent with the template schema | ✅ int 0–800, matching templates |

## C. images / versions

| Spec item | 3.1 (schema) | v8 install (WIP) |
|---|---|---|
| `md5sum` renamed `checksum`, old name still accepted | ✅ both accepted | ✅ |
| `checksum_type` optional, MD5 only for now | ✅ enum contains only `md5` | ✅ |
| `compression_target` alongside `compression` | ✅ | ✅ |
| `versions[].images` (hda…/cdrom…) | ✅ | ✅ merged into the template; IOU maps `image` → `path` |
| version-level `category`, `usage`, `symbol`, `default_username`, `default_password`, `installation_instructions` | ✅ | ✅ version-level values override appliance-level ones in metadata |
| `emulator_versions` (struck from the spec) | ✅ correctly absent | ✅ |
| installed version injected into the template `version` field (controller >= 3.0) | ❌ no v8 conversion at all | ✅ |

## D. Install path (behavioral requirements)

| Spec item | 3.1 | v8 install (WIP) |
|---|---|---|
| A v8 appliance installs into a template | ❌ `new_template()` only understands the v1-6 layout (see repro above) | ✅ full v8 conversion path (`_new_template_v8`), covered by ~600 lines of conversion tests |
| v2.2.41+/3.0 both support v8 | — (deployment-level) | — |

## Summary

- **Stock 3.1** implements the *schema* side of the spec almost completely (the one hard violation being the required `kvm`), but the **install/conversion side is missing entirely** — which is exactly why v8 appliances currently list but cannot install.
- **The WIP install branch** covers every spec item, with two lenient supersets (`kvm` still accepted as optional, version-level `idlepc` fallback) and a few extensions beyond the spec (`netmiko_device_type`, docker `custom_adapters` / `extra_configs`, appliance metadata persisted on the template).
